### ISSUE TITLE
Clarifying the Role of Hyperbolic Layers in Hypformer

### ISSUE THREAD

**Issue by AIyumeng** [Graph-and-Geometric-Learning/hyperbolic-transformer#7]
In the referenced work, the Hypformer is constructed from a series of hyperbolic layers. However, these layers actually separate the spatial and temporal components:

the spatial transformations are performed in Euclidean space,

and the temporal component is obtained through a deterministic function rather than a learned temporal module.

My question is:
If the main operations are still carried out in Euclidean space and the temporal information is provided by an explicit function, what fundamentally distinguishes the Hypformer from the Eucformer? Why is it still considered a Hypformer?

```python
class HypLayerNorm(nn.Module):
    def __init__(self, manifold, in_features, manifold_out=None):
        super(HypLayerNorm, self).__init__()
        self.in_features = in_features
        self.manifold = manifold
        self.manifold_out = manifold_out
        self.layer = nn.LayerNorm(self.in_features)
        self.reset_parameters()

    def reset_parameters(self):
        self.layer.reset_parameters()

    def forward(self, x):
        x_space = x[..., 1:]
        x_space = self.layer(x_space)
        x_time = ((x_space**2).sum(dim=-1, keepdims=True) + self.manifold.k).sqrt()
        x = torch.cat([x_time, x_space], dim=-1)

        if self.manifold_out is not None:
            x = x * (self.manifold_out.k / self.manifold.k).sqrt()
        return x

class HypNormalization(nn.Module):
    def __init__(self, manifold, manifold_out=None):
        super(HypNormalization, self).__init__()
        self.manifold = manifold
        self.manifold_out = manifold_out

    def forward(self, x):
        x_space = x[..., 1:]
        x_space = x_space / x_space.norm(dim=-1, keepdim=True)
        x_time = ((x_space**2).sum(dim=-1, keepdims=True) + self.manifold.k).sqrt()
        x = torch.cat([x_time, x_space], dim=-1)
        if self.manifold_out is not None:
            x = x * (self.manifold_out.k / self.manifold.k).sqrt()
        return x

class HypActivation(nn.Module):
    def __init__(self, manifold, activation, manifold_out=None):
        super(HypActivation, self).__init__()
        self.manifold = manifold
        self.manifold_out = manifold_out
        self.activation = activation

    def forward(self, x):
        x_space = x[..., 1:]
        x_space = self.activation(x_space)
        x_time = ((x_space**2).sum(dim=-1, keepdims=True) + self.manifold.k).sqrt()
        x = torch.cat([x_time, x_space], dim=-1)
        if self.manifold_out is not None:
            x = x * (self.manifold_out.k / self.manifold.k).sqrt()
        return x

class HypDropout(nn.Module):
    def __init__(self, manifold, dropout, manifold_out=None):
        super(HypDropout, self).__init__()
        self.manifold = manifold
        self.manifold_out = manifold_out
        self.dropout = nn.Dropout(dropout)

    def forward(self, x, training=False):
        if training:
            x_space = x[..., 1:]
            x_space = self.dropout(x_space)
            x_time = ((x_space**2).sum(dim=-1, keepdims=True) + self.manifold.k).sqrt()
            x = torch.cat([x_time, x_space], dim=-1)
            if self.manifold_out is not None:
                x = x * (self.manifold_out.k / self.manifold.k).sqrt()
        return x

class HypLinear(nn.Module):
    """
    Parameters:
        manifold (manifold): The manifold to use for the linear transformation.
        in_features (int): The size of each input sample.
        out_features (int): The size of each output sample.
        bias (bool, optional): If set to False, the layer will not learn an additive bias. Default is True.
        dropout (float, optional): The dropout probability. Default is 0.1.
    """

    def __init__(self, manifold, in_features, out_features, bias=True, manifold_out=None):
        super().__init__()
        self.in_features = in_features + 1  # + 1 for time dimension
        self.out_features = out_features
        self.bias = bias
        self.manifold = manifold
        self.manifold_out = manifold_out
        
        self.linear = nn.Linear(self.in_features, self.out_features, bias=bias)
        self.reset_parameters()

    def reset_parameters(self):
        init.xavier_uniform_(self.linear.weight, gain=math.sqrt(2))
        init.constant_(self.linear.bias, 0)

    def forward(self, x, x_manifold='hyp'):
        if x_manifold != 'hyp':
            x = torch.cat([torch.ones_like(x)[..., 0:1], x], dim=-1)
            x = self.manifold.expmap0(x)
        x_space = self.linear(x)

        x_time = ((x_space**2).sum(dim=-1, keepdims=True) + self.manifold.k).sqrt()
        x = torch.cat([x_time, x_space], dim=-1)
        if self.manifold_out is not None:
            x = x * (self.manifold_out.k / self.manifold.k).sqrt()
        return x
```

**Comment by marlin-codes** [comment]
Thanks for your question! Let me address this systematically.

While it's true that the spatial operations appear to use Euclidean matrices $W$, the crucial point is **where and how these transformations act**. Let me write out the transformations explicitly:

**Pseudo-Lorentz Rotation** (transformation without time and space interaction):
$$(\sqrt{\|Wx_{space}\|^2-1/K}, Wx_{space}) = \begin{pmatrix} \frac{\sqrt{\|Wx_{space}\|^2 - 1/K}}{x_{time}} & 0 \\ 0 & W \end{pmatrix} \begin{pmatrix} x_{time} \\ x_{space} \end{pmatrix}$$

**Pseudo-Lorentz Boost** (transformation on both time and space-like dimensions):
$$(\sqrt{\|Wx\|^2-1/K}, Wx) = \begin{pmatrix} \sqrt{\|Wx\|^2 - 1/K} \cdot e_{0,:} & v_{0,:} \\ \sqrt{\|Wx\|^2 - 1/K} \cdot e_{1,:} & W_{1,:,:} \end{pmatrix} \begin{pmatrix} x_{time} \\ x_{space} \end{pmatrix}$$

**Your Core Concern: "Spatial transformations in Euclidean space"**

Your observation is astute, but there's a subtle yet critical point: **the spatial transformations are NOT actually performed in Euclidean space**, they're performed on the spatial components of points that live in hyperbolic space. This distinction is fundamental.

**1. The transformation is fundamentally $f: \mathbb{H}^{n,K} \to \mathbb{H}^{n,K}$**

The key insight is that even though matrix $W$ appears "Euclidean," the complete transformation operates on and returns points on the hyperboloid:

- **Input**: $x \in \mathbb{H}^{n,K} = \{x \in \mathbb{R}^{n+1} : -x_0^2 + \sum_{i=1}^n x_i^2 = -1/K, x_0 > 0\}$
- **Output**: $x' \in \mathbb{H}^{n,K}$ (still satisfies the Lorentzian constraint)
- **Metric preserved**: Hyperbolic distance $d_{\mathbb{H}}(x,y) = \frac{1}{\sqrt{K}} \text{arccosh}(-K\langle x,y \rangle_L)$

The matrix $W$ doesn't operate independently — it's part of a larger transformation that maintains hyperbolic geometry. This is fundamentally different from Euclidean transformations where $x' = Wx$ and that's it.

**2. Why the "Deterministic Temporal Function" is Essential, Not a Limitation**

You note that "the temporal component is obtained through a deterministic function rather than a learned temporal module." This is precisely what makes it a valid hyperbolic transformation! Here's why:

**In Euclidean Space (Eucformer)**:
- No constraint between components
- Each dimension can be transformed independently
- $x' = Wx$ is a complete, valid transformation

**In Hyperbolic Space (Hypformer)**:
- The hyperboloid constraint $\|x\|^2_L = -x_0^2 + \|x_{space}\|^2 = -1/K$ must be satisfied at all times
- If you transform the spatial components: $x'_{space} = Wx_{space}$
- The temporal component **cannot be arbitrary** — it must be: $x'_{time} = \sqrt{\|x'_{space}\|^2 - 1/K}$
- This isn't "avoiding learning" — it's **enforcing the geometric structure**

Think of it this way: if you walk on the surface of a sphere and move in the $x$ and $y$ directions, your $z$ coordinate is determined by $x^2 + y^2 + z^2 = r^2$. You don't "learn" the $z$ coordinate separately — it's constrained by the geometry. The same principle applies here with the Lorentzian constraint.

**3. The Pseudo-Lorentz Structure Changes Everything**

Let me clarify why these transformations are fundamentally hyperbolic:

**Pseudo-Lorentz Rotation**:
$$x'_{time} = \sqrt{\|Wx_{space}\|^2 - 1/K}$$
$$x'_{space} = Wx_{space}$$

Even though $W$ only acts on $x_{space}$, this is a **hyperbolic transformation** because:
- The temporal component is recomputed to maintain $-(x'_{time})^2 + \|x'_{space}\|^2 = -1/K$
- The Lorentzian distance between points is preserved under this transformation
- The transformation corresponds to a rotation within the hyperbolic space (geodesic-preserving)

**Pseudo-Lorentz Boost**:
$$x'_{time} = \sqrt{\|Wx\|^2 - 1/K} \cdot e_0 \cdot x_{time} + v_0 \cdot x_{space}$$
$$x'_{space} = \sqrt{\|Wx\|^2 - 1/K} \cdot e_1 \cdot x_{time} + W_1 \cdot x_{space}$$

Here, the off-diagonal terms $(v_0, e_0, e_1)$ explicitly couple time and space, creating hyperbolic translations that have no Euclidean counterpart. These parameters are **learned** and perform genuine hyperbolic geometric operations.

**4. What Fundamentally Distinguishes Hypformer from Eucformer**

Let me directly address your core question with a concrete comparison:

**Eucformer operates in $\mathbb{R}^n$**:
- Transformation: $x' = Wx$
- Metric: $\langle x, y \rangle = x^T y$ (Euclidean)
- Distance: $d(x,y) = \|x - y\|_2$
- No constraints between components
- Learned parameters: $W \in \mathbb{R}^{d \times d}$

**Hypformer operates in $\mathbb{H}^{n,K}$**:
- Transformation: $x' = (\sqrt{\|Wx_{space}\|^2 - 1/K}, Wx_{space})$ or full boost
- Metric: $\langle x, y \rangle_L = -x_0 y_0 + \sum_{i=1}^n x_i y_i$ (Lorentzian)
- Distance: $d(x,y) = \frac{1}{\sqrt{K}} \text{arccosh}(-K\langle x,y \rangle_L)$
- Constraint: $\|x\|^2_L = -1/K$ enforced via deterministic temporal component
- Learned parameters: $W \in \mathbb{R}^{d \times d}$, but interpreted through Lorentzian geometry; additionally $(v_0, e_0, e_1, W_1)$ for boosts

The fundamental difference is NOT in the appearance of $W$, but in:
1. **The space where points live**: $\mathbb{H}^{n,K}$ vs $\mathbb{R}^n$
2. **The metric structure**: Lorentzian vs Euclidean
3. **The constraint surface**: Hyperboloid vs unconstrained
4. **The induced geometry**: Constant negative curvature vs flat

**5. Why "Spatial Operations" Are Not Really "Euclidean Operations"**

This is the crux of the confusion. When we write $Wx_{space}$, it **looks** like a Euclidean operation, but it's actually:

- Operating on the spatial components of a hyperbolic point
- Part of a larger map $f: \mathbb{H}^{n,K} \to \mathbb{H}^{n,K}$
- Constrained to preserve the Lorentzian structure
- Measured by hyperbolic distance, not Euclidean distance

Consider an analogy: when you perform rotations on a sphere using matrices, you're using "Euclidean-looking" rotation matrices, but the operation is still fundamentally spherical geometry because points stay on the sphere and distances are measured along the sphere's surface.

Similarly, even though $W$ looks Euclidean, the complete transformation:
$$x \mapsto (\sqrt{\|Wx_{space}\|^2 - 1/K}, Wx_{space})$$
is a **hyperbolic isometry** (or more generally, a hyperbolic diffeomorphism), not a Euclidean transformation.
